File: java-code/src/main/java/refactoring/refactoring_api/remove_argument_flag/after/Order.java
package refactoring.refactoring_api.remove_argument_flag.after;


import java.time.LocalDate;
import java.util.Arrays;

public class Order {

  String deliveryState;
  LocalDate placedOn;

    public static void main(String[] args) {
      var order = new Order();

      //STEP 2: Call methods without argument flag
      var rushDeliveryDate = order.rushDeliveryDate();
      var regularDeliveryDate = order.regularDeliveryDate();
    }

    //STEP 1: Apply decompose conditional
    //STEP 3: I dont need deliveryDate method anymore
    public LocalDate deliveryDate(boolean isRush) {
      if (isRush) {
        return rushDeliveryDate();
      } else {
        return regularDeliveryDate();
      }
    }

  public LocalDate regularDeliveryDate() {
    int deliveryTime;
    if (Arrays.asList("MA", "CT", "NY").contains(this.deliveryState)) {
      deliveryTime = 2;
    } else if (Arrays.asList("ME", "NH").contains(this.deliveryState)) {
      deliveryTime = 3;
    } else {
      deliveryTime = 4;
    }
    return this.placedOn.plusDays(2 + deliveryTime);
  }

  public LocalDate rushDeliveryDate() {
    int deliveryTime;
    if (Arrays.asList("MA", "CT").contains(this.deliveryState)) {
      deliveryTime = 1;
    } else if (Arrays.asList("NY", "NH").contains(this.deliveryState)) {
      deliveryTime = 2;
    } else {
      deliveryTime = 3;
    }
    return this.placedOn.plusDays(1 + deliveryTime);
  }
}

File: go-code/refactoring_api/remove_flag_argument/after/order.go
package after

import (
	"log"
	"time"
)

type Order struct {
	DeliveryState string
	PlacedOn      time.Time
}

//STEP 1 : I refactor the method by Decompose Conditional
//STEP 3: I dont need deliveryDate func anymore
/*func deliveryDate(anOrder Order, isRush bool) time.Time {
	if isRush {
		return rushDeliveryData(anOrder)
	} else {
		return regularDeliveryDate(anOrder)
	}
}*/

func regularDeliveryDate(anOrder Order) time.Time {
	var deliveryTime int

	if contains([]string{"MA", "CT", "NY"}, anOrder.DeliveryState) {
		deliveryTime = 2
	} else if contains([]string{"ME", "NH"}, anOrder.DeliveryState) {
		deliveryTime = 3
	} else {
		deliveryTime = 4
	}
	return anOrder.PlacedOn.AddDate(0, 0, 2+deliveryTime)
}

func rushDeliveryData(anOrder Order) time.Time {
	var deliveryTime int

	if contains([]string{"MA", "CT"}, anOrder.DeliveryState) {
		deliveryTime = 1
	} else if contains([]string{"NY", "NH"}, anOrder.DeliveryState) {
		deliveryTime = 2
	} else {
		deliveryTime = 3
	}
	return anOrder.PlacedOn.AddDate(0, 0, 1+deliveryTime)
}

func contains(slice []string, target string) bool {
	for _, item := range slice {
		if item == target {
			return true
		}
	}
	return false
}

func main() {
	order := Order{
		DeliveryState: "MA",
		PlacedOn:      time.Now(),
	}

	//STEP 2: I can now call directly the function that doesnt require argument flag
	notRushDeliveryDate := regularDeliveryDate(order)
	rushDeliveryDate := rushDeliveryData(order)

	log.Printf("notRushDeliveryDate = %v\n", notRushDeliveryDate)
	log.Printf("rushDeliveryDate = %v\n", rushDeliveryDate)
}
